File: swim-system-java/swim-core-java/swim.structure/src/main/java/swim/structure/Extant.java
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.structure;

import swim.codec.Output;
import swim.util.Murmur3;

public final class Extant extends Value {

  private Extant() {
    // singleton
  }

  /**
   * Always returns {@code true} because {@code Extant} is a defined value.
   */
  @Override
  public boolean isDefined() {
    return true;
  }

  /**
   * Always returns {@code false} because {@code Extant} is not a distinct
   * value.
   */
  @Override
  public boolean isDistinct() {
    return false;
  }

  /**
   * Always returns {@code false} because {@code Extant} is not a definite value.
   */
  @Override
  public boolean isDefinite() {
    return false;
  }

  @Override
  public boolean isConstant() {
    return true;
  }

  /**
   * Always returns an empty {@code Record} because {@code Extant} is not a
   * distinct value.
   */
  @Override
  public Record unflattened() {
    return Record.empty();
  }

  @Override
  public Value not() {
    return Value.absent();
  }

  /**
   * Always returns the empty {@code String} because {@code Extant} behaves
   * like an empty {@code Record}, which converts to a {@code String} by
   * concatenating the string values of all its members, if all its members
   * convert to string values.
   */
  @Override
  public String stringValue() {
    return "";
  }

  /**
   * Always returns the empty {@code String} because {@code Extant} behaves
   * like an empty {@code Record}, which converts to a {@code String} by
   * concatenating the string values of all its members, if all its members
   * convert to string values.
   */
  @Override
  public String stringValue(String orElse) {
    return "";
  }

  /**
   * Always returns {@code true} because {@code Extant} behaves like a truthy
   * value.
   */
  @Override
  public boolean booleanValue() {
    return true;
  }

  /**
   * Always returns {@code true} because {@code Extant} behaves like a truthy
   * value.
   */
  @Override
  public boolean booleanValue(boolean orElse) {
    return true;
  }

  @Override
  public int typeOrder() {
    return 98;
  }

  @Override
  public int compareTo(Item other) {
    return Integer.compare(this.typeOrder(), other.typeOrder());
  }

  @Override
  public boolean equals(Object other) {
    return this == other;
  }

  private static int hashSeed;

  @Override
  public int hashCode() {
    if (Extant.hashSeed == 0) {
      Extant.hashSeed = Murmur3.seed(Extant.class);
    }
    return Extant.hashSeed;
  }

  @Override
  public <T> Output<T> debug(Output<T> output) {
    output = output.write("Value").write('.').write("extant").write('(').write(')');
    return output;
  }

  private static final Extant VALUE = new Extant();

  public static Extant extant() {
    return Extant.VALUE;
  }

}

File: swim-js/swim-runtime/swim-core/swim-structure/src/main/Extant.ts
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Lazy, Numbers, Constructors, Interpolator} from "@swim/util";
import type {Output} from "@swim/codec";
import {Item} from "./Item";
import {Value} from "./Value";
import {Record} from "./Record";

/** @public */
export type AnyExtant = Extant | null;

/** @public */
export class Extant extends Value {
  /** @internal */
  private constructor() {
    super();
  }

  /**
   * Always returns `true` because `Extant` is a defined value.
   */
  override isDefined(): boolean {
    return true;
  }

  /**
   * Always returns `false` because `Extant` is not a distinct value.
   */
  override isDistinct(): boolean {
    return false;
  }

  /**
   * Always returns `false` because `Extant` is not a definite value.
   */
  override isDefinite(): boolean {
    return false;
  }

  override isConstant(): boolean {
    return true;
  }

  /**
   * Always returns an empty `Record` because `Extant` is not a distinct value.
   */
  override unflattened(): Record {
    return Record.empty();
  }

  override not(): Value {
    return Value.absent();
  }

  /**
   * Always returns the empty `string` because `Extant` behaves like an empty
   * `Record`, which converts to a `string` by concatenating the string values
   * of all its members, if all its members convert to string values.
   */
  override stringValue(): string;
  /**
   * Always returns the empty `string` because `Extant` behaves like an empty
   * `Record`, which converts to a `string` by concatenating the string values
   * of all its members, if all its members convert to string values.
   */
  override stringValue<T>(orElse: T): string;
  override stringValue<T>(orElse?: T): string {
    return "";
  }

  /**
   * Always returns `true` because `Extant` behaves like a truthy value.
   */
  override booleanValue(): boolean;
  /**
   * Always returns `true` because `Extant` behaves like a truthy value.
   */
  override booleanValue<T>(orElse: T): boolean;
  override booleanValue<T>(orElse?: T): boolean {
    return true;
  }

  override toAny(): AnyExtant {
    return null;
  }

  override interpolateTo(that: Extant): Interpolator<Extant>;
  override interpolateTo(that: Item): Interpolator<Item>;
  override interpolateTo(that: unknown): Interpolator<Item> | null;
  override interpolateTo(that: unknown): Interpolator<Item> | null {
    return super.interpolateTo(that);
  }

  override get typeOrder(): number {
    return 98;
  }

  override compareTo(that: unknown): number {
    if (that instanceof Item) {
      return Numbers.compare(this.typeOrder, that.typeOrder);
    }
    return NaN;
  }

  override equivalentTo(that: unknown): boolean {
    return this === that;
  }

  override equals(that: unknown): boolean {
    return this === that;
  }

  override hashCode(): number {
    return Constructors.hash(Extant);
  }

  override debug<T>(output: Output<T>): Output<T> {
    output = output.write("Value").write(46/*'.'*/).write("extant").write(40/*'('*/).write(41/*')'*/);
    return output;
  }

  override display<T>(output: Output<T>): Output<T> {
    output = output.write("null");
    return output;
  }

  @Lazy
  static override extant(): Extant {
    return new Extant();
  }

  static override fromAny(value: AnyExtant): Extant {
    if (value instanceof Extant) {
      return value;
    } else if (value === null) {
      return Extant.extant();
    } else {
      throw new TypeError("" + value);
    }
  }
}
